### PR TITLE
[DataGrid] Fix support for `@material-ui/core@4.12`

### LINKED ISSUE #2106: [DataGrid] Broken support of core v4.12.1
in data-grid version 4.0.0-alpha.33, i cannot build my typescript project because the data-grid.d.ts:519 references Localization$1['components'] in @material-ui/core/localization, which does not exist.
I have tried with the listed version @material-ui/core@4.9.12 and the latest version @material-ui/core@4.12.1.
The problem is also in version 4.0.0-alpha.32 but not present in 4.0.0-alpha.31 as it does not reference the localization of material ui.

The error message:
Property 'components' does not exist on type 'Localization'.

Maybe it is meant to be only for material-ui@5.0.0? i havent tried with that, as it is not tagged as the latest version yet.

### PR DESCRIPTION
**Pull request by DanailH**
Fixes #2106

A full explanation of the problem can be found here -> https://github.com/mui-org/material-ui-x/issues/2106#issuecomment-876996807

- v4.11.4: https://codesandbox.io/s/material-demo-forked-dtst2?file=/demo.js
- v4.12.1: https://codesandbox.io/s/material-demo-forked-skokv?file=/demo.js
- v5.0.0.beta-0: https://codesandbox.io/s/material-demo-forked-n0qqf?file=/demo.js

### CODE REVIEW

File: packages/grid/_modules_/grid/utils/utils.ts [hunk 1]
```diff
@@ -22,7 +22,8 @@ export function getThemePaletteMode(palette: any): string {
 }
 
 export function isMuiV5(): boolean {
-  return 'alpha' in styles;
+  // Changed from `alpha` to 'fade' because mui/core@4.12.0 adds 'alpha' in v4
```
Comment (flaviendelangle) [inline packages/grid/_modules_/grid/utils/utils.ts:25]: Could we check several properties to reduce the risk of a future break ?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/utils/utils.ts:25]: Nice trick but this looks like a comment for a the PR's description, not the source

```suggestion
```

**Review by oliviertassinari — DISMISSED** [review]
"Update isMuiV5 helper" assuming what we go after is "Fix support for @material-ui/core@4.12", we also need to update `muiStyleAlpha` and `createTheme` to remove the deprecation warnings. We can no longer test the version, we need to feature detect.

**Comment by DanailH** [comment]
@oliviertassinari Yes, I had a quick chat with @mnajdova this morning about it that was the easiest solution to the problem.

**Comment by oliviertassinari** [comment]
> The drawback with doing that is that we need to also update the core version we support. Basically to use the grid one will need to use core@4.12.0 > , no ?

No need, we can feature detect.

File: packages/grid/_modules_/grid/utils/utils.ts [hunk 2]
```diff
@@ -22,18 +22,19 @@ export function getThemePaletteMode(palette: any): string {
 }
 
 export function isMuiV5(): boolean {
-  return 'alpha' in styles;
+  return !('fade' in styles);
 }
 
 export function muiStyleAlpha(color: string, value: number): string {
-  if (isMuiV5()) {
+  if (isMuiV5() || !(styles as any)?.fade) {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/utils/utils.ts:29]: Why not?

```suggestion
  if ((styles as any)?.alpha) {
```

x the other case

Comment (DanailH) [reply packages/grid/_modules_/grid/utils/utils.ts:29]: I wanted to still utilize the `isMuiV5` helper but in those two cases, it doesn't make much sense.

We would still need to keep it tho, `isMuiV5` as it is used on other places as well.

**Comment by DanailH** [comment]
The tests are failing because of depreciation warnings. Basically now we need a way to know if it is v4.12 because I don't see how we can detect things like `<TablePagination onChangePage />` or `<TablePagination onPageChange />`

**Comment by oliviertassinari** [comment]
@DanailH Arf, maybe we should do what you have proposed initially then, drop <v4.12 versions support.

File: packages/grid/_modules_/grid/utils/utils.ts [hunk 3]
```diff
@@ -21,21 +21,31 @@ export function getThemePaletteMode(palette: any): string {
...
+  if ('fade' in styles && 'alpha' in styles) {
+    return 'v4.12';
+  }
+
+  return 'v4';
 }
 
 export function muiStyleAlpha(color: string, value: number): string {
-  if (isMuiV5()) {
+  if ((styles as any)?.alpha) {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/utils/utils.ts:37]: A good lesson for us, feature detection >> version matching.

**Comment by bartvde** [comment]
any chance there will be a release soon that has this fix incorporated? We are running into: 

```Warning: Failed prop type: Invalid prop `modifiers` of type `array` supplied to `ForwardRef(Popper)`, expected `object`.``` 

because of this one. TIA.

**Comment by striky1** [comment]
> any chance there will be a release soon that has this fix incorporated? We are running into:
> 
> `` Warning: Failed prop type: Invalid prop `modifiers` of type `array` supplied to `ForwardRef(Popper)`, expected `object`. ``
> 
> because of this one. TIA.

Same here + I still have errors:

```
index.js:1 Material-UI: The key `selectLabel` provided to the classes prop is not implemented in ForwardRef(TablePagination).
You can only override one of the following: root,toolbar,spacer,caption,selectRoot,select,selectIcon,input,menuItem,actions.
```

I tried downgrade back of `/core` to `4.11.4` but it didn't help.

Current package.json:

```
    "@material-ui/core": "^4.12.2",
    "@material-ui/data-grid": "^4.0.0-alpha.33",
    "@material-ui/icons": "^4.11.2",
    "@material-ui/lab": "^4.0.0-alpha.58",
    "@material-ui/pickers": "^3.3.10",
    "@material-ui/styles": "^4.11.4",
```

Thank you

**Comment by Primajin** [comment]
```
    "@material-ui/core": "^4.12.2",
    "@material-ui/icons": "^4.11.2",
    "@material-ui/lab": "4.0.0-alpha.60",
    "@material-ui/pickers": "^3.3.10",
    "@material-ui/x-grid": "^4.0.0-alpha.34",
```
seems to have resolved the issue

**Comment by striky1** [comment]
> ```
>     "@material-ui/core": "^4.12.2",
>     "@material-ui/icons": "^4.11.2",
>     "@material-ui/lab": "4.0.0-alpha.60",
>     "@material-ui/pickers": "^3.3.10",
>     "@material-ui/x-grid": "^4.0.0-alpha.34",
> ```
> 
> 
>      
>   
> 
> seems to have resolved the issue

but you are using `x-grid` instead of `data-grid` what is not MIT version

**Comment by oliviertassinari** [comment]
@striky1 He's a Pro plan user. This fix was released in MUI X version: 4.0.0-alpha.34, regardless of the plan (MIT or commercial).

**Pull request merged**